### PR TITLE
Add train from scratch function

### PR DESCRIPTION
**Pull request by irinakhismatullina**
Train from scratch function, just wraps up all preparation functions.

I am not shure of the best way to handle parameters here. All the preparation functions will also have wrappings for CL usage. So I prefer passing dictionaries to specifying all the defaults in several places, but I'm not shure it's a good practice.
Will be glad to any advice on the subject.

### CODE REVIEW

File: lookout/style/typos/preparation.py [hunk 1]
```diff
@@ -40,6 +29,26 @@
     "raw_data_filename": "raw_data.csv",
     "vocabulary_filename": "vocabulary.csv",
     "frequencies_filename": "frequencies.csv",
+    "prepared_filename": "prepared.csv",
+}
+
+defaults_for_fasttext = {
+    "size": 100000000,
```
Comment (vmarkovtsev) [inline lookout/style/typos/preparation.py]: Would be nice to have a comment here what does this huge size mean

File: lookout/style/typos/preparation.py [hunk 2]
```diff
@@ -40,6 +29,26 @@
...
+    "corrupt": True,
+    "typo_probability": 0.2,
+    "add_typo_probability": 0.005,
+    "fasttext_path": str(DATA_DIR / "emb.bin"),
```
Comment (vmarkovtsev) [inline lookout/style/typos/preparation.py]: Better to use long names: embeddings.bin

And btw what is `bin`? If it is the numpy format, we should use `npy`, etc.

File: lookout/style/typos/preparation.py [hunk 3]
```diff
@@ -84,14 +93,14 @@ def prepare_data(params: Optional[Mapping[str, Any]] = None) -> pandas.DataFrame
                                      frequencies for all present tokens will be saved.
                    raw_data_filename: Name of .csv file in data_dir to put raw dataset in case of \
                                       loading from drive.
-                   vocabulary_path: Name of .csv file in data_dir to save vocabulary to.
-                   frequencies_path: Name of .csv file in data_dir to save frequencies to.
+                   vocabulary_filename: Name of .csv file in data_dir to save vocabulary to.
```
Comment (vmarkovtsev) [inline lookout/style/typos/preparation.py]:
```suggestion
                   vocabulary_filename: Name of the .csv file in `data_dir` to save vocabulary to.
```

(the same below)

File: lookout/style/typos/preparation.py [hunk 4]
```diff
@@ -200,9 +183,53 @@ def train_fasttext(data: pandas.DataFrame, params: Optional[Mapping[str, Any]] =
...
+        test.to_csv(params["test_path"])
+    if params["train_path"] is not None:
+        train.to_csv(params["train_path"])
+    return train, test
+
+
 def train_and_evaluate(train_data: pandas.DataFrame, test_data: pandas.DataFrame,
                        vocabulary_path: str, frequencies_path: str, fasttext_path: str,
-                       threads_number: int = 8) -> TyposCorrector:
+                       threads_number: int = 16) -> TyposCorrector:
```
Comment (vmarkovtsev) [inline lookout/style/typos/preparation.py]: You should never hardcode these values. `multiprocessing.cpu_count()`

File: lookout/style/typos/preparation.py [hunk 5]
```diff
@@ -225,3 +252,46 @@ def train_and_evaluate(train_data: pandas.DataFrame, test_data: pandas.DataFrame
...
     model.evaluate(test_data)
     return model
+
+
+def train_from_scratch(prepare_params: Optional[Mapping[str, Any]] = None,
+                       pretrained_fasttext: Optional[str] = None,
+                       fasttext_params: Optional[Mapping[str, Any]] = None,
+                       datasets_params: Optional[Mapping[str, Any]] = None,
+                       save_model_path: Optional[str] = None,
+                       threads_number: int = 16) -> TyposCorrector:
```
Comment (vmarkovtsev) [inline lookout/style/typos/preparation.py]: same here - `threads_number`

File: lookout/style/typos/preparation.py (hunk 2 above)
Comment (irinakhismatullina) [reply lookout/style/typos/preparation.py]: It's fasttext default format for saving full model.

File: lookout/style/typos/preparation.py [hunk 6]
```diff
@@ -200,9 +185,54 @@ def train_fasttext(data: pandas.DataFrame, params: Optional[Mapping[str, Any]] =
...
+                          Columns.Split, Columns.Frequency and Columns.Token.
+    :param params: Parameters for creating train and test datasets, options:
+                   train_size: Train dataset size.
+                   test_size: Test dataset size.
+                   typo_probability: Probability with which a token gets to be corrupted.
+                   add_typo_probability: Probability with which one more corruption happens to a \
+                                         corrupted token.
+                   train_path: Path to the .csv file where to save the train dataset.
+                   test_path: Path to the .csv file where to save the test dataset.
+    :param threads_number: Number of threads for multiprocessing.
```
Comment (vmarkovtsev) [inline lookout/style/typos/preparation.py]: multiprocessing works wit htprocesses, not with threads. Those two are very different, especially for Python. So do you mean `processes_number` here?

Comment (irinakhismatullina) [reply lookout/style/typos/preparation.py]: Yes, it's processes, changing it everywhere.

**Pull request merged**